### PR TITLE
Feat/delete ghost neighbours

### PR DESCRIPTION
**Pull request by mastricker**
Only a draft PR, but possible to comment on, especially if we messed something up in the tests.

This PR gets rid of the previous notion of _ghost atoms_. As discussed in the meeting we now only have two _types_ of atoms: centers and ghosts (i.e. not-centers, this includes masked center atoms now see issue #180).
The change has been reflected in how `Properties` are sized, etc.

Further, this PR changes the Neighbourlist algorithm so that ghost atoms do not have neighbours anymore. This is also reflected in the `AdaptorMaxOrder` (see issue #155). We started with the introduction of the `IsCompactCluster` flag to give the child a name. The non-compact-cluster would be the centered one (where all distances of the atoms in the new Cluster are checked to be small than the cutoff).But that is part of a different PR.

Further: AdaptorMaxOrder now only works on the minimal list (see #31).

Rationale and detailed description of the changes:

- Neighbours for ghost atoms are not build anymore
- Flag for `Representations` to size the `Property` to _only_ the number of centers.
- AdaptorMaxOrder has `IsCompactCluster` (centred to follow).
- Somewhat overlapping with the error catching mechanism @felixmusil opened (PR #196), but his version is more detailed and will be more helpful.

I am not filling out the things below before feedback.

Tasks before review:

- [x] Add tests, examples, and complete documentation of any added functionality
    - [x] Make sure the autogenerated documentation appears in Sphinx and is
          formatted correctly (ask @max-veit if you need help with this task).
    - [x] Write additional documentation in the Sphinx (not docstrings!) to
          explain the feature and its usage in plain English
    - [x] Make sure the examples run (!) and produce the expected output
    - [x] For bugfix pull requests, list here which tests have been added or re-enabled
          to verify the fix and catch future regressions as well as similar bugs
          elsewhere in the code.
- [X] Run `make lint` on the project, ensure it passes
    - [X] Run `make pretty-cpp` and `make pretty-python`, check that the
          auto-formatting is sensible
    - [X] Review variable names, make sure they're descriptive (ask a friend)
    - [X] Review all TODOs, convert to issues wherever possible
    - [X] Make sure draft, in-progress, and commented-out code is moved to its
          own branch
- [x] If committing any code in Jupyter/IPython notebooks, install and run `nbstripout`
- [x] Merge with master and resolve any conflicts
- [x] (anything else that's still in progress)

Remaining tasks, out of scope of this pull request: (optional)

-
-
- [x] Make sure to add these points to the issues or to the meeting agenda so they don't
get forgotten!

### CODE REVIEW

File: scripts/generate_kernel_ref_data.py [hunk 1]
```diff
@@ -1,3 +1,8 @@
+import rascal
+import rascal.lib as lrl
+from rascal.representations import SphericalInvariants
+from rascal.models import Kernel
+from rascal.utils import ostream_redirect
```
[the remark this answers is not in the document]
Comment (mastricker) [reply scripts/generate_kernel_ref_data.py]: It is weird, I did not touch those. I guess they were reformatted during the `pretty-python` run. Same as the rest below

File: src/structure_managers/property.hh [hunk 2]
```diff
@@ -81,10 +81,20 @@ namespace rascal {
     Property() = delete;
 
     //! Constructor with Manager
+    template <size_t MyOrder = Order,
+              std::enable_if_t<MyOrder != 1, void *> = nullptr>
```
[the remark this answers is not in the document]
Comment (mastricker) [reply src/structure_managers/property.hh]: There are other places in property, where the void pointer is also used / was used before. Seemed to be coherent when we programmed it.

File: src/structure_managers/property_block_sparse.hh [hunk 3]
```diff
@@ -746,6 +730,101 @@ namespace rascal {
...
+                        std::string metadata = "no metadata")
+        : Parent{manager, metadata} {}
+
+    virtual ~BlockSparseProperty() = default;
+
+    //! Adjust size of values (only increases, never frees)
+    inline void resize() final {
+      size_t new_size{this->template get_validated_property_length<Order>()};
+      this->values.resize(new_size);
+    }
```
[the remark this answers is not in the document]
Comment (mastricker) [reply src/structure_managers/property_block_sparse.hh]: I don't really know how. This way it is explicit?

File: src/representations/calculator_sorted_coulomb.hh [hunk 4]
```diff
@@ -426,22 +426,22 @@ namespace rascal {
 
     // Get a reference to the data container where the computed representation
     // is stored in the corresponding structure manager
-    auto && coulomb_matrices{
-        manager->template get_property_ref<Prop_t>(this->get_name())};
+    constexpr bool ExcludeGhosts{true};
+    auto coulomb_matrices{manager->template get_property_ptr<Prop_t>(
```
[the remark this answers is not in the document]
Comment (mastricker) [reply src/representations/calculator_sorted_coulomb.hh]: They both make the same in the end. We probably should decide on the simpler version and only have one?

File: src/structure_managers/adaptor_neighbour_list.hh [hunk 5]
```diff
@@ -545,16 +531,16 @@ namespace rascal {
...
     inline size_t get_nb_clusters(size_t order) const {
       switch (order) {
-      case 1: {
-        return this->get_size_with_ghosts();
-        break;
-      }
+      // case 1: {
+      //   return this->get_size_with_ghosts();
+      //   break;
+      // }
```
[the remark this answers is not in the document]
Comment (mastricker) [reply src/structure_managers/adaptor_neighbour_list.hh]: It does not make sense here any more. `get_nb_cluster` is/was only used in resizing. But we now have an ambiguity for clusters of order 1 - does the user want with or without ghosts. So why no explicitly call the size/size_with_ghosts. Should go away everywhere.

File: src/structure_managers/property_block_sparse.hh (hunk 3 above)
[the remark this answers is not in the document]
Comment (felixmusil) [reply src/structure_managers/property_block_sparse.hh]: let's discuss this next meeting

[the remark this answers is not in the document]
Comment (mastricker) [reply src/structure_managers/property_block_sparse.hh]: I'll put it in the _minutes_

File: examples/json_structure.cc [hunk 6]
```diff
@@ -93,7 +92,7 @@ int main() {
   manager->update(structure);
 
   std::cout << "Number of centers: " << manager->get_size() << std::endl;
-  std::cout << "Number of atoms (total): " << manager->get_n_atoms()
+  std::cout << "Number of atoms (total): " << manager->get_size_with_ghosts()
```
[the remark this answers is not in the document]
Comment (mastricker) [reply examples/json_structure.cc]: I changed the output

File: src/representations/calculator_sorted_coulomb.hh (hunk 4 above)
[the remark this answers is not in the document]
Comment (felixmusil) [reply src/representations/calculator_sorted_coulomb.hh]: makes sense. let's keep get_property_ptr and de-reference explicitly at that point ?

**Comment by mastricker** [comment]
The thing with the logic about the resizing in my opinion is: If we go the proposed route to have it default to the centers+ghosts, we are on the safe side. So you explicitly have to ask for a smaller Property, if you know what you are doing, i.e. you want to optimize space.
But if you fail to optimize, the worst thing that happens is that you have a large property. Meaning nothing can go wrong, it is just a waste of space.
And the only point where we explicitly ask for a smaller Property is in the representations. If a developer looks at existing RepresentationCalculators, he/she will notice the peculiarity.

Thats why I would go with the larger default.

File: src/structure_managers/property.hh (hunk 2 above)
[the remark this answers is not in the document]
Comment (felixmusil) [reply src/structure_managers/property.hh]: in the rest of the code int is preferred to void*

[the remark this answers is not in the document]
Comment (mastricker) [reply src/structure_managers/property.hh]: I have no strong feeling. Seems to me that the lines that Till added are done with the nullptr and the ones you or someone else added are done with the int version. I don't care as long as it is consistent. -> meeting minutes?
Maybe there are some subtleties I am not aware of...

File: src/structure_managers/adaptor_strict.hh [hunk 7]
```diff
@@ -453,6 +451,19 @@ namespace rascal {
       }
     }
 
+    for (auto && atom : this->manager->only_ghosts()) {
```
[the remark this answers is not in the document]
Comment (Luthaf) [reply src/structure_managers/adaptor_strict.hh]: > the variables that are instantiated like that are often unreachable in the debugger.

That's strange. This works with no issue for me (gcc + lldb) 

```cpp
#include <string>
#include <vector>

struct Test {
  Test(std::string name, int count) : name(std::move(name)), count(count) {}
  Test(Test &&) = default;
  Test(const Test &) = delete;
  Test &operator=(Test &&) = default;
  Test &operator=(const Test &) = delete;

  std::string name;
  int count;
};

int main() {
  std::vector<Test> values;
  values.emplace_back("foo", 40);
  values.emplace_back("bar", -33);

  int count = 0;
  for (auto &&value : values) {
    count += value.count;
  }

  return 0;
}
```

```
$ g++ -g debug-auto-move.cpp -o debug-auto-move
$ lldb ./debug-auto-move
(lldb) target create "./debug-auto-move"
Current executable set to './debug-auto-move' (x86_64).
(lldb) breakpoint set -f debug-auto-move.cpp -l 22
Breakpoint 1: where = debug-auto-move`main + 182 at debug-auto-move.cpp:22, address = 0x0000000000000cc0
(lldb) r
Process 4371 launched: './debug-auto-move' (x86_64)
Process 4371 stopped
* thread #1, name = 'debug-auto-move', stop reason = breakpoint 1.1
    frame #0: 0x0000555555554cc0 debug-auto-move`main at debug-auto-move.cpp:22
   19  	
   20  	  int count = 0;
   21  	  for (auto &&value : values) {
-> 22  	    count += value.count;
   23  	  }
   24  	
   25  	  return 0;
(lldb) print value
(Test) $0 = (name = std::__cxx11::string @ 0x000055a9722db350, count = 40)
(lldb) continue
Process 4371 resuming
Process 4371 stopped
* thread #1, name = 'debug-auto-move', stop reason = breakpoint 1.1
    frame #0: 0x0000555555554cc0 debug-auto-move`main at debug-auto-move.cpp:22
   19  	
   20  	  int count = 0;
   21  	  for (auto &&value : values) {
-> 22  	    count += value.count;
   23  	  }
   24  	
   25  	  return 0;
(lldb) print value
(Test) $1 = (name = std::__cxx11::string @ 0x000055a97255e150, count = -33)
```

---

Maybe the problem is with the CLion debugger you are using?

[the remark this answers is not in the document]
Comment (felixmusil) [reply src/structure_managers/adaptor_strict.hh]: it uses gdb which is also fairly standard

**Comment by mastricker** [comment]
I included the changes we talked about in the meeting.

Then I rebased and locally tests pass.
Hope I did not mess something up here.

This is good for review.

In essence: got rid of the `consider_ghost_neighbour` flag with all including changes.

**Comment by felixmusil** [comment]
Seems like you undid quite a few of the changes introduced recently in master with this merge, e.g. move of src/ to src/rascal.

**Comment by mastricker** [comment]
Well I'm on it. Before your merges there was not conflict.

**Comment by mastricker** [comment]
This merge is a nightmare. I am rebasing onto stuff that I already rebased onto and reviewing changes that I already accepted with the previous merge. Will take a while.

Can we in the future please please please go forward with merge requests which are done from the review side and were rebase/merged with the current state of master. This is really painful and could have been avoided.

**Comment by felixmusil** [comment]
I am not too sure I understand. What do you mean ?

**Comment by mastricker** [comment]
That I am rebasing onto master, but the changes that are presented and have to be resolved are more or less the same as when I rebased the last time. I don't know why git thinks the already accepted changes are still _new_.

**Comment by Luthaf** [comment]
> That I am rebasing onto master, but the changes that are presented and have to be resolved are more or less the same as when I rebased the last time. I don't know why git thinks the already accepted changes are still new.

I think this problem is related to the weird rebase/merge hybrid you created last time. 

[`git rerere`](https://git-scm.com/book/en/v2/Git-Tools-Rerere) can help if you are fixing the same conflicts over and over (I never used it myself, I just know it exist)

**Comment by mastricker** [comment]
I think a rebase will not work. I did another rebase and it completely disregarded the change in folder.
I will try a merge...

**Comment by felixmusil** [comment]
Ok so the branch is no again up to date with master and I also in removed the intermediate base class for properties for the exclude_ghosts member.

File: docs/source/neighbour_list/adaptors.rst [hunk 8]
```diff
@@ -17,21 +17,25 @@ A typical usage scenario is an :class:`~AdaptorStrict` on top of an :class:`~Ada
 
 Each adaptor only stores the additional information it generates during construction. No atoms or neighbourlists are stored twice. If the requested information does not exist at this `Layer`, it is piped to the underlying adaptor/manager.
 
+Within librascal we decided to distinguish between two cases for atoms. An atom can be treated as a **center** or as a **ghost**. These two terms are ambiguous because they are used differently in different contexts and they are important for some consideration of what is possible with this library. Before going into the details of adapting atomic structures to a specific use case, we therefore define the two groups:
+
+1. **Ghost** atoms are all atoms which are either masked out or only neighbours, sometimes termed *j-atoms*. These include the newly generated periodic images in the :class:`~AdaptorNeighbourList`.
+2. **Center** atoms are atoms which are to be worked, i.e. *i-atoms*. Th
```
Comment (felixmusil) [inline docs/source/neighbour_list/adaptors.rst]:
```suggestion
2. **Center** atoms are atoms which are to be worked, i.e. *i-atoms*.
```

File: src/rascal/structure_managers/adaptor_filter.hh [hunk 9]
```diff
@@ -184,32 +184,37 @@ namespace rascal {
...
      * return the number of 'neighbours' (i.e., number of pairs for an atom,
      * number of triplets for a pair, etc) of a given order.
+     *
+     * Note: The case for order=1 is abmiguous: one possible answer is the
+     * number of centers the other possibility is the number of centers +
+     * ghost atoms. Please use the get_size or get_size_with_ghosts member
+     * functions.
      */
-    size_t get_nb_clusters(int order) const {
+    inline size_t get_nb_clusters(int order) const {
```
Comment (felixmusil) [inline src/rascal/structure_managers/adaptor_filter.hh]: are the inline statements really needed in this file ?

File: src/rascal/structure_managers/adaptor_strict.hh [hunk 10]
```diff
@@ -136,14 +138,17 @@ namespace rascal {
     //! returns the (strict) cutoff for the adaptor
     double get_cutoff() const { return this->cutoff; }
 
-    size_t get_nb_clusters(int order) const {
-      assert(order > 0);
+    inline size_t get_nb_clusters(int order) const {
+      assert(order > 1);
+      assert(order <= 2);
```
Comment (felixmusil) [inline src/rascal/structure_managers/adaptor_strict.hh]: why not assert(order == 2) ?

**Comment by mastricker** [comment]
Is that a memory error on gcc-9-debug?
```
g++-9: fatal error: Killed signal terminated program cc1plus
compilation terminated.
performance/profiles/CMakeFiles/profile_spherical_expansion.dir/build.make:62: recipe for target 'performance/profiles/CMakeFiles/profile_spherical_expansion.dir/profile_spherical_expansion.cc.o' failed
make[2]: *** [performance/profiles/CMakeFiles/profile_spherical_expansion.dir/profile_spherical_expansion.cc.o] Error 1
CMakeFiles/Makefile2:2101: recipe for target 'performance/profiles/CMakeFiles/profile_spherical_expansion.dir/all' failed
make[1]: *** [performance/profiles/CMakeFiles/profile_spherical_expansion.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 47%] Building CXX object bindings/CMakeFiles/_rascal.dir/bind_py_utils.cc.o
[ 48%] Building CXX object bindings/CMakeFiles/_rascal.dir/bind_py_representation_calculator.cc.o
[ 50%] Building CXX object bindings/CMakeFiles/_rascal.dir/bind_py_models.cc.o
[ 50%] Linking CXX shared module _rascal.cpython-36m-x86_64-linux-gnu.so
[ 50%] Built target _rascal
Makefile:140: recipe for target 'all' failed
make: *** [all] Error 2
Exited with code 2
```
Local compilation works flawlessly.

**Comment by mastricker** [comment]
Ok, retriggering helped. Can we make a merge please @felixmusil , pretty please?

**Comment by mastricker** [comment]
Ok, HOPEFULLY the last merge with master for now...

**Pull request merged**